### ISSUE TITLE
what is the mnist network's activation function and  backward loss?

### ISSUE THREAD

**Issue by LebronX** [NeuralNetworkVerification/Marabou#448]
Hi,

I am trying to reproduce the result from your FMCAD paper. I would like to know what is the mnist network's activation function and backward loss. I couldn't find the saved model for mnist or the code for describing the architecture. 

Thanks

**Comment by anwu1219** [comment]
Hi,

We evaluated on all-ReLU networks. The MNIST networks we evaluated are contained here:
https://github.com/NeuralNetworkVerification/Marabou/tree/master/resources/nnet/mnist

In particular we evaluated on MNIST10x10, MNIST10x20 and MNIST20x20.
10x10 has 10 layers, 10 ReLUs per layer.
10x20 has 10 layers and 20 ReLU per layer.
20x20 has 20 layers and 20 ReLUs per layer.

A description of the .nnet format can be found here: https://github.com/sisl/NNet

And you could use this script to generate targeted attack queries (I believe we describe the epsilon values that we tested on in the paper):
https://github.com/NeuralNetworkVerification/Marabou/blob/master/resources/properties/mnist/dump_mnist_targeted_attack.py

The following folder contains some example properties generated by the script above:
https://github.com/NeuralNetworkVerification/Marabou/tree/master/resources/properties/mnist

Please let me know if you have other questions!

Andrew

Andrew

**Comment by LebronX** [comment]
Hi,

Thanks for the answer! 

If you also use ReLU in the last layer (instead of the usual softmax), I would like to ask what lost function do you use to compute the loss?

In addition, in your 10x20 and 20x20 version, how do you decide the prediction of the network? The usual architecture use argmax to take the biggest value among 10 output neurons. But how do you decide this in 20 neurons setting?

Thanks!

**Comment by anwu1219** [comment]
Hi,

There is no activation functions in the output layer. That is, it is just an affine transformation of the outputs of the last hidden layer.
At training time, we did apply softmax to the output layer, and we used the canonical Cross Entropy Loss.
At verification time, we remove the softmax layer, as it does not change the classification result. (i.e., max(softmax(y)) = max(y), where y is the pre-softmax neural network output).

Andrew

**Comment by LebronX** [comment]
Hi,

Thank you very much for the explanation! Now I understand.

One more question: How do you "remove" the softmax at verification time? It looks like for reading ".onnx" file, it will raise "NotImplementedError" when reading softmax.

Best

**Comment by anwu1219** [comment]
Hi, I think it's possible to not add the softmax layer when you define the network architecture at training time.
Both tensorflow's `softmax_cross_entropy_with_logits` and pytorch's `CrossEntropyLoss` methods apply softmax to the network output.

**Comment by LebronX** [comment]
Thank you very much for the answering!
